clang/test/CodeGenSYCL/device_global.cpp

```diff
@@ -0,0 +1,103 @@
+
+// decorated with only global_variable_allowed attribute
+template <typename T>
+class [[__sycl_detail__::global_variable_allowed]] only_global_var_allowed {
+public:
+  const T &get() const noexcept { return *Data; }
+  only_global_var_allowed() {}
+  operator T &() noexcept { return *Data; }
+
+private:
+  T *Data;
+};
+
+// check that we don't generate `sycl-unique-id` IR attribute if class does not use
+// [[__sycl_detail__::device_global]]
+only_global_var_allowed<int> no_device_global;
+// CHECK: @no_device_global = addrspace(1) global %class.only_global_var_allowed zeroinitializer, align 8{{$}}
+
+inline namespace Bar {
+device_global<float> InlineNS;
+}
+// CHECK: @_ZN3Bar8InlineNSE = addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global.0" zeroinitializer, align 8 #[[BAR_INLINENS_ATTRS:[0-9]+]]
+
+template <typename T> struct TS {
+public:
+  static device_global<T> d;
+};
+template <> device_global<int> TS<int>::d{};
+// CHECK: @_ZN2TSIiE1dE = addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[TEMPLATED_WRAPPER_ATTRS:[0-9]+]]
+
+template <typename T>
+device_global<T> templ_dev_global;
+// CHECK: @[[TEMPL_DEV_GLOB:[a-zA-Z0-9_]+]] = linkonce_odr addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, comdat, align 8 #[[TEMPL_DEV_GLOB_ATTRS:[0-9]+]]
+
+void foo() {
+  q.submit([&](handler &h) {
+    h.single_task<class kernel_name_1>([=]() {
+      (void)A;
+      (void)B;
+      (void)Foo::C;
+      (void)same_name;
+      (void)NS::same_name;
+      (void)no_device_global;
+      (void)Bar::InlineNS;
+      auto AA = TS<int>::d.get();
+      auto val = templ_dev_global<int>.get();
+    });
+  });
+}
+
+namespace {
+device_global<int> same_name;
+}
+// CHECK: @_ZN12_GLOBAL__N_19same_nameE = internal addrspace(1) global %"class.cl::sycl::ext::oneapi::device_global" zeroinitializer, align 8 #[[SAME_NAME_ANON_NS_ATTRS:[0-9]+]]
+
+namespace {
+void bar() {
+  q.submit([&](handler &h) {
+    h.single_task<class kernel_name>([=]() { int A = same_name; });
+  });
+}
+} // namespace
+
+// CHECK: @llvm.global_ctors = appending global [2 x { i32, void ()*, i8 addrspace(4)* }] [{ i32, void ()*, i8 addrspace(4)* } { i32 65535, void ()* @__cxx_global_var_init{{.*}}, i8 addrspace(4)* addrspacecast (i8 addrspace(1)* bitcast (%"class.cl::sycl::ext::oneapi::device_global" addrspace(1)* @[[TEMPL_DEV_GLOB]] to i8 addrspace(1)*) to i8 addrspace(4)*) }, { i32, void ()*, i8 addrspace(4)* } { i32 65535, void ()* @_GLOBAL__sub_I_device_global.cpp, i8 addrspace(4)* null }]
+// CHECK: @llvm.used = appending global [1 x i8 addrspace(4)*] [i8 addrspace(4)* addrspacecast (i8 addrspace(1)* bitcast (%"class.cl::sycl::ext::oneapi::device_global" addrspace(1)* @[[TEMPL_DEV_GLOB]] to i8 addrspace(1)*) to i8 addrspace(4)*)], section "llvm.metadata"
```

Comment:
https://github.com/intel/llvm/pull/5597/commits/5bced1a651bb78c46754af0c26a75a006950232f added some address space related changes to emission of llvm.global_ctors and llvm.used .
It was done to fix the crash inside clang codegen (on attempt to emit bitcast between two pointers of different address space) and complains from llvm-spirv about address space casts between disjoint address spaces.
It seems when variable template is used (the one like `templ_dev_global` in this test), unlike the others it is referenced directly from `llvm.global.ctors` and `llvm.used` arrays. These arrays contain pointers inside. In LLVM IR emitted for SYCL with spir-based triple global variables have global (1) address space. I believe no-one have ever tried to use real global variables that require `llvm.global.ctors` using targets that use address spaces, so these were emitted with no (private) address space and therefore when a global variable was referenced from there - incorrect casts were emitted. I was following the general logic of address space support in DPC++ and added default address space to pointers inside `llvm.global_ctors` and `llvm.used` arrays.  Now it generates LLVM IR that is considered as valid by the SPIR-V translator.

Reply:
Thanks for this explanation @Fznamznon